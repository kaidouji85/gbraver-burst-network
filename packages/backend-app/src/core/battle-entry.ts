import type { ArmdozerId, PilotId } from "gbraver-burst-core";

import type { UserID } from "./user";
import type { WSAPIGatewayConnectionId } from "./ws-api-gateway-connection";

/**
 * ネット対戦エントリ
 * カジュアルマッチ、プライベートマッチ等、全ケースのエントリで
 * 必要最小限の情報をあつめた
 */
export type BattleEntry = {
  /** エントリーするユーザのID */
  userID: UserID;

  /** 選択したアームドーザID */
  armdozerId: ArmdozerId;

  /** 選択したパイロットID */
  pilotId: PilotId;

  /** コネクションID */
  connectionId: WSAPIGatewayConnectionId;
};
