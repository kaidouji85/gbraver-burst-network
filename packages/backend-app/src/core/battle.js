// @flow

import type { Player } from "gbraver-burst-core";
import type { GameState } from "gbraver-burst-core/lib/state/game-state";

import type { UserID } from "./user";
import type { WSAPIGatewayConnectionId } from "./ws-api-gateway-connection";

/** バトルID */
export type BattleID = string;

/** ステートが更新されるたびに発行されるユニークID */
export type FlowID = string;

/** バトルに参加するプレイヤー情報 */
export type BattlePlayer = Player & {
  /** ユーザID */
  userID: UserID,
  /** コネクションID */
  connectionId: WSAPIGatewayConnectionId,
};

/**
 * バトル情報
 * @template X プレイヤー情報
 */
export type Battle<X: BattlePlayer> = {
  /** バトルID */
  battleID: BattleID,
  /** フローID */
  flowID: FlowID,
  /** プレイヤー情報 */
  players: [X, X],
  /**
   * バトル更新ポーリングをするユーザのID
   * playersに含まれているユーザのIDを指定すること
   */
  poller: UserID,
  /** ステートヒストリー */
  stateHistory: GameState[],
};
