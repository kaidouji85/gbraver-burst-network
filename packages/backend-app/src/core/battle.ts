import { GameState, Player, PlayerSchema } from "gbraver-burst-core";
import { z } from "zod";

import { UserID, UserIDSchema } from "./user";
import {
  WSAPIGatewayConnectionId,
  WSAPIGatewayConnectionIdSchema,
} from "./ws-api-gateway-connection";

/** バトルID */
export type BattleID = string;

/** バトルID zodスキーマ */
export const BattleIDSchema = z.string();

/** ステートが更新されるたびに発行されるユニークID */
export type FlowID = string;

/** FLowID zodスキーマ */
export const FlowIDSchema = z.string();

/** バトルに参加するプレイヤー情報 */
export type BattlePlayer = Player & {
  /** ユーザID */
  userID: UserID;
  /** コネクションID */
  connectionId: WSAPIGatewayConnectionId;
};

/** BattlePlayer zodスキーマ */
export const BattlePlayerSchema = PlayerSchema.extend({
  userID: UserIDSchema,
  connectionId: WSAPIGatewayConnectionIdSchema,
});

/**
 * バトル情報
 * @template X プレイヤー情報
 */
export type Battle<X extends BattlePlayer> = {
  /** バトルID */
  battleID: BattleID;
  /** フローID */
  flowID: FlowID;
  /** プレイヤー情報 */
  players: [X, X];
  /**
   * バトル更新ポーリングをするユーザのID
   * playersに含まれているユーザのIDを指定すること
   */
  poller: UserID;
  /** ステートヒストリー */
  stateHistory: GameState[];
};
