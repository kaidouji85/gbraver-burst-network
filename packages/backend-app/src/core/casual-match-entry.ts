import { BattleEntry, BattleEntrySchema } from "./battle-entry";

/** カジュアルマッチエントリ */
export type CasualMatchEntry = BattleEntry;

/** CasualMatchEntry zodスキーマ */
export const CasualMatchEntrySchema = BattleEntrySchema;
