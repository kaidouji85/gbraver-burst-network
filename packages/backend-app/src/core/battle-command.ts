import { Command, CommandSchema } from "gbraver-burst-core";
import { z } from "zod";

import { BattleID, BattleIDSchema, FlowID, FlowIDSchema } from "./battle";
import { UserID, UserIDSchema } from "./user";

/** バトルコマンド */
export type BattleCommand = {
  /** ユーザID */
  userID: UserID;
  /** バトルID */
  battleID: BattleID;
  /** フローID */
  flowID: FlowID;
  /** コマンド */
  command: Command;
};

/** BattleCommand zodスキーマ */
export const BattleCommandSchema = z.object({
  userID: UserIDSchema,
  battleID: BattleIDSchema,
  flowID: FlowIDSchema,
  command: CommandSchema,
});