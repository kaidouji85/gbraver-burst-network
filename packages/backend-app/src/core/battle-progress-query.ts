import { BattleID, FlowID } from "./battle";

/** バトル進行 問合わせパラメータ */
export type BattleProgressQuery = {
  /** バトルID */
  battleID: BattleID;
  /** フローID */
  flowID: FlowID;
};