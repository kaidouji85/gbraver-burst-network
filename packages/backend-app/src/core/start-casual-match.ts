import { Battle, BattlePlayer } from "./battle";
import { CasualMatchEntry } from "./casual-match-entry";
import { CasualMatching } from "./casual-match-make";
import { Connection } from "./connection";
import { createBattle } from "./create-battle";
import { createBattlePlayer } from "./create-battle-player";

/** カジュアルマッチ開始情報 */
type MatchStartResponse = {
  /** バトル情報 */
  battle: Battle<BattlePlayer>;
  /** バトル参加者のコネクション */
  connections: Connection[];
}

/**
 * カジュアルマッチを開始する
 * @param matching マッチング
 * @return バトルとコネクションのリスト
 */
export function startCasualMatch<X extends CasualMatchEntry> (
  matching: CasualMatching<X>,
): MatchStartResponse {
  const battle = createBattle([
    createBattlePlayer(matching[0]),
    createBattlePlayer(matching[1]),
  ]);
  const connections: Connection[] = matching.map((v) => ({
    connectionId: v.connectionId,
    userID: v.userID,
    state: {
      type: "InBattle",
      battleID: battle.battleID,
      players: battle.players,
    },
  }));
  return { battle, connections };
}