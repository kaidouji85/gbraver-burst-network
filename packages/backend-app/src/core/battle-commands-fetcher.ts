import { BattlePlayer } from "./battle";
import { BattleCommand } from "./battle-command";

/** ゲーム参加プレイヤーのバトルコマンドを取得 */
export interface BattleCommandsFetcher {
  /**
   * ゲーム参加プレイヤーのバトルコマンドを取得する
   * @param players ゲーム参加プレイヤー
   * @return 取得結果、取得できなかった場合はnull
   */
  fetch(
    players: [BattlePlayer, BattlePlayer],
  ): Promise<[BattleCommand, BattleCommand] | null>;
}
