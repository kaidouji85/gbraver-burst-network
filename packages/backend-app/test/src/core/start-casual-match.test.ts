import { ArmdozerIds, PilotIds } from "gbraver-burst-core";
import { v4 as uuidv4 } from "uuid";

import { CasualMatchEntry } from "../../../src/core/casual-match-entry";
import { startCasualMatch } from "../../../src/core/start-casual-match";
import { mockUniqUUID } from "../../mock-unique-uuid";

jest.mock("uuid");

/** カジュアルマッチ参加プレイヤー1 */
const entry01: CasualMatchEntry = {
  userID: "user-01",
  armdozerId: ArmdozerIds.SHIN_BRAVER,
  pilotId: PilotIds.SHINYA,
  connectionId: "user-01-connection-id",
};

/** カジュアルマッチ参加プレイヤー2 */
const entry02: CasualMatchEntry = {
  userID: "user-02",
  armdozerId: ArmdozerIds.NEO_LANDOZER,
  pilotId: PilotIds.GAI,
  connectionId: "user-02-connection-id",
};

beforeEach(() => {
  (uuidv4 as jest.Mock).mockImplementation(mockUniqUUID());
});

afterEach(() => {
  (uuidv4 as jest.Mock).mockReset();
});

test("カジュアルマッチを正しく開始できる", () => {
  expect(startCasualMatch([entry01, entry02])).toMatchSnapshot();
});
