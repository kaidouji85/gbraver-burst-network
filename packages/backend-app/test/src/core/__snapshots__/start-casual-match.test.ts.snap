// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`カジュアルマッチを正しく開始できる 1`] = `
{
  "battle": {
    "battleID": "mock-id-3",
    "flowID": "mock-id-4",
    "players": [
      {
        "armdozer": {
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "id": "SHIN_BRAVER",
          "maxBattery": 5,
          "maxHp": 3100,
          "name": "シンブレイバー",
          "power": 2000,
          "speed": 2000,
        },
        "connectionId": "user-01-connection-id",
        "pilot": {
          "id": "SHINYA",
          "name": "シンヤ",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "mock-id-1",
        "userID": "user-01",
      },
      {
        "armdozer": {
          "burst": {
            "buffPower": 1000,
            "duration": 2,
            "recoverBattery": 3,
            "type": "BuffPower",
          },
          "id": "NEO_LANDOZER",
          "maxBattery": 5,
          "maxHp": 3300,
          "name": "ネオランドーザ",
          "power": 2300,
          "speed": 1500,
        },
        "connectionId": "user-02-connection-id",
        "pilot": {
          "id": "GAI",
          "name": "ガイ",
          "skill": {
            "buffPower": 600,
            "duration": 2,
            "type": "BuffPowerSkill",
          },
        },
        "playerId": "mock-id-2",
        "userID": "user-02",
      },
    ],
    "poller": "user-01",
    "stateHistory": [
      {
        "activePlayerId": "mock-id-1",
        "effect": {
          "name": "StartGame",
        },
        "players": [
          {
            "armdozer": {
              "battery": 5,
              "burst": {
                "recoverBattery": 5,
                "turnStartBatteryCorrect": 1,
                "type": "RecoverBattery",
              },
              "effects": [],
              "enableBurst": true,
              "hp": 3100,
              "id": "SHIN_BRAVER",
              "maxBattery": 5,
              "maxHp": 3100,
              "name": "シンブレイバー",
              "power": 2000,
              "speed": 2000,
            },
            "pilot": {
              "enableSkill": true,
              "id": "SHINYA",
              "name": "シンヤ",
              "skill": {
                "recoverBattery": 2,
                "type": "RecoverBatterySkill",
              },
            },
            "playerId": "mock-id-1",
          },
          {
            "armdozer": {
              "battery": 5,
              "burst": {
                "buffPower": 1000,
                "duration": 2,
                "recoverBattery": 3,
                "type": "BuffPower",
              },
              "effects": [],
              "enableBurst": true,
              "hp": 3300,
              "id": "NEO_LANDOZER",
              "maxBattery": 5,
              "maxHp": 3300,
              "name": "ネオランドーザ",
              "power": 2300,
              "speed": 1500,
            },
            "pilot": {
              "enableSkill": true,
              "id": "GAI",
              "name": "ガイ",
              "skill": {
                "buffPower": 600,
                "duration": 2,
                "type": "BuffPowerSkill",
              },
            },
            "playerId": "mock-id-2",
          },
        ],
      },
      {
        "activePlayerId": "mock-id-1",
        "effect": {
          "name": "InputCommand",
          "players": [
            {
              "command": [
                {
                  "battery": 0,
                  "type": "BATTERY_COMMAND",
                },
                {
                  "battery": 1,
                  "type": "BATTERY_COMMAND",
                },
                {
                  "battery": 2,
                  "type": "BATTERY_COMMAND",
                },
                {
                  "battery": 3,
                  "type": "BATTERY_COMMAND",
                },
                {
                  "battery": 4,
                  "type": "BATTERY_COMMAND",
                },
                {
                  "battery": 5,
                  "type": "BATTERY_COMMAND",
                },
                {
                  "type": "BURST_COMMAND",
                },
                {
                  "type": "PILOT_SKILL_COMMAND",
                },
              ],
              "playerId": "mock-id-1",
              "selectable": true,
            },
            {
              "command": [
                {
                  "battery": 0,
                  "type": "BATTERY_COMMAND",
                },
                {
                  "battery": 1,
                  "type": "BATTERY_COMMAND",
                },
                {
                  "battery": 2,
                  "type": "BATTERY_COMMAND",
                },
                {
                  "battery": 3,
                  "type": "BATTERY_COMMAND",
                },
                {
                  "battery": 4,
                  "type": "BATTERY_COMMAND",
                },
                {
                  "battery": 5,
                  "type": "BATTERY_COMMAND",
                },
                {
                  "type": "BURST_COMMAND",
                },
                {
                  "type": "PILOT_SKILL_COMMAND",
                },
              ],
              "playerId": "mock-id-2",
              "selectable": true,
            },
          ],
        },
        "players": [
          {
            "armdozer": {
              "battery": 5,
              "burst": {
                "recoverBattery": 5,
                "turnStartBatteryCorrect": 1,
                "type": "RecoverBattery",
              },
              "effects": [],
              "enableBurst": true,
              "hp": 3100,
              "id": "SHIN_BRAVER",
              "maxBattery": 5,
              "maxHp": 3100,
              "name": "シンブレイバー",
              "power": 2000,
              "speed": 2000,
            },
            "pilot": {
              "enableSkill": true,
              "id": "SHINYA",
              "name": "シンヤ",
              "skill": {
                "recoverBattery": 2,
                "type": "RecoverBatterySkill",
              },
            },
            "playerId": "mock-id-1",
          },
          {
            "armdozer": {
              "battery": 5,
              "burst": {
                "buffPower": 1000,
                "duration": 2,
                "recoverBattery": 3,
                "type": "BuffPower",
              },
              "effects": [],
              "enableBurst": true,
              "hp": 3300,
              "id": "NEO_LANDOZER",
              "maxBattery": 5,
              "maxHp": 3300,
              "name": "ネオランドーザ",
              "power": 2300,
              "speed": 1500,
            },
            "pilot": {
              "enableSkill": true,
              "id": "GAI",
              "name": "ガイ",
              "skill": {
                "buffPower": 600,
                "duration": 2,
                "type": "BuffPowerSkill",
              },
            },
            "playerId": "mock-id-2",
          },
        ],
      },
    ],
  },
  "connections": [
    {
      "connectionId": "user-01-connection-id",
      "state": {
        "battleID": "mock-id-3",
        "players": [
          {
            "armdozer": {
              "burst": {
                "recoverBattery": 5,
                "turnStartBatteryCorrect": 1,
                "type": "RecoverBattery",
              },
              "id": "SHIN_BRAVER",
              "maxBattery": 5,
              "maxHp": 3100,
              "name": "シンブレイバー",
              "power": 2000,
              "speed": 2000,
            },
            "connectionId": "user-01-connection-id",
            "pilot": {
              "id": "SHINYA",
              "name": "シンヤ",
              "skill": {
                "recoverBattery": 2,
                "type": "RecoverBatterySkill",
              },
            },
            "playerId": "mock-id-1",
            "userID": "user-01",
          },
          {
            "armdozer": {
              "burst": {
                "buffPower": 1000,
                "duration": 2,
                "recoverBattery": 3,
                "type": "BuffPower",
              },
              "id": "NEO_LANDOZER",
              "maxBattery": 5,
              "maxHp": 3300,
              "name": "ネオランドーザ",
              "power": 2300,
              "speed": 1500,
            },
            "connectionId": "user-02-connection-id",
            "pilot": {
              "id": "GAI",
              "name": "ガイ",
              "skill": {
                "buffPower": 600,
                "duration": 2,
                "type": "BuffPowerSkill",
              },
            },
            "playerId": "mock-id-2",
            "userID": "user-02",
          },
        ],
        "type": "InBattle",
      },
      "userID": "user-01",
    },
    {
      "connectionId": "user-02-connection-id",
      "state": {
        "battleID": "mock-id-3",
        "players": [
          {
            "armdozer": {
              "burst": {
                "recoverBattery": 5,
                "turnStartBatteryCorrect": 1,
                "type": "RecoverBattery",
              },
              "id": "SHIN_BRAVER",
              "maxBattery": 5,
              "maxHp": 3100,
              "name": "シンブレイバー",
              "power": 2000,
              "speed": 2000,
            },
            "connectionId": "user-01-connection-id",
            "pilot": {
              "id": "SHINYA",
              "name": "シンヤ",
              "skill": {
                "recoverBattery": 2,
                "type": "RecoverBatterySkill",
              },
            },
            "playerId": "mock-id-1",
            "userID": "user-01",
          },
          {
            "armdozer": {
              "burst": {
                "buffPower": 1000,
                "duration": 2,
                "recoverBattery": 3,
                "type": "BuffPower",
              },
              "id": "NEO_LANDOZER",
              "maxBattery": 5,
              "maxHp": 3300,
              "name": "ネオランドーザ",
              "power": 2300,
              "speed": 1500,
            },
            "connectionId": "user-02-connection-id",
            "pilot": {
              "id": "GAI",
              "name": "ガイ",
              "skill": {
                "buffPower": 600,
                "duration": 2,
                "type": "BuffPowerSkill",
              },
            },
            "playerId": "mock-id-2",
            "userID": "user-02",
          },
        ],
        "type": "InBattle",
      },
      "userID": "user-02",
    },
  ],
}
`;
